### DISCUSSION TITLE
Microsoft.Data.SqlClient not working.

### DISCUSSION THREAD

**Discussion by mateli** [pythonnet/pythonnet#2307] [Q&A]
### Environment

-   Pythonnet version: 3.0.3
-   Python version: 3.12.0
-   Operating System: Windows 11
-   .NET Runtime: 8.0.0

### Details

from pythonnet import load
load("coreclr")
import clr
clr.AddReference(r"Microsoft.Data.SqlClient.dll")
from Microsoft.Data.SqlClient import SqlConnection
try:
    conn = SqlConnection("")
except Exception as e:
    print(e)
    print (System.Environment.StackTrace)
    exit()

Output:
Microsoft.Data.SqlClient is not supported on this platform.
   at Microsoft.Data.SqlClient.SqlConnection..ctor(String connectionString)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeConstructorWithoutAlloc(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Environment.get_StackTrace()
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
   at System.Reflection.RuntimePropertyInfo.GetValue(Object obj, Object[] index)
   at Python.Runtime.PropertyObject.tp_descr_get(BorrowedReference ds, BorrowedReference ob, BorrowedReference tp) in /tmp/build-via-sdist-h3j67_wc/pythonnet-3.0.3/src/runtime/Types/PropertyObject.cs:line 65

More info:
Various online sources suggest that failure to find "Microsoft.Data.SqlClient.SNI.dll" (native part of the implementation) is the root cause of this failure. However putting this dll in the path or manually loading it seems to make no difference. Either it is manually loaded by SqlClient in a weird way or something else is causing this error.

Another suggested solution is to use a very old version of SqlClient or the System.Data.SqlClient and neither are good options as they are obsolete and unreliable. 

Also adding a non-empty connection string doesn't matter because SqlClient do not seem to get to the point where this is processed.

**Comment by geographika** [comment]
Ran into this also. Through trial and error, I copied DLLs from a build output in the following order:

1. Copy all DLLs from `\bin\Release\net8.0` (this included `Microsoft.Data.SqlClient.dll`).

    Error `System.PlatformNotSupportedException: Microsoft.Data.SqlClient is not supported on this platform.`

2. Copy Microsoft.Data.SqlClient.dll and the other DLLs from following location and replace the existing one from `\bin\Release\net8.0\runtimes\win\lib\net8.0`

    Error: `System.DllNotFoundException: Unable to load DLL 'Microsoft.Data.SqlClient.SNI.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)`

3. Copy Microsoft.Data.SqlClient.SNI.dll from: `bin\Release\net8.0\runtimes\win-x64\native`

Worked!

**Reply by mateli** [reply]
You copied them where?

**Reply by geographika** [reply]
I had a DLL with a reference to `Microsoft.Data.SqlClient.dll`, and had the same error as you. I wasn't trying to use `Microsoft.Data.SqlClient.dll` directly though. I copied my DLL and all dependent ones into a single folder and added that to `sys.path`. 

```
import sys
from pythonnet import load
load("coreclr")
import clr
sys.path.append("/folder/with/dlls") # My.dll and all dependent DLLs
clr.AddReference("My.dll")
```

![image](https://github.com/pythonnet/pythonnet/assets/490840/e2f69f8c-628c-43e6-b17f-8c1440b32e99)

**Reply by vfofanov** [reply]
@geographika Thank you a lot! You save my day))

I prepared MSBuild targets file [PrepareRuntimes.targets](https://github.com/user-attachments/files/18142964/PrepareRuntimes.txt) for automate this operation. This `.targets` specific for `net8` or `net9` and `Microsoft.Data.SqlClient` Version=`5.2.2`. The difference will be in `runtimes` subfolders

#### How to use it

1. Add the attached file near your `.csproj`
2. Add `Import` to your `.csproj`(somewhere inside `<Project>`)
```xml
<Import Project="$(MSBuildThisFileDirectory)..\PrepareRuntimes.targets" />
```
3. Build 
4. You will find `runtime-specific` folder in output folder
5. For get runtime information in Python I'm using the next script(checked only on windows x64)
```python
import platform


def get_runtime():
    system = platform.system().lower()  # e.g., 'windows', 'linux', 'darwin'
    architecture = platform.machine().lower()  # e.g., 'x86_64', 'aarch64'

    # based on https://stackoverflow.com/questions/45125516/possible-values-for-uname-m
    # based on https://en.wikipedia.org/wiki/Uname
    # Mapping of system and architecture to custom platform names
    if system in ['windows', 'windows_nt'] and architecture == 'amd64':
        return 'win-x64'
    elif system == 'linux' and architecture in ['arm', 'arm64', 'aarch64_b', 'aarch64', 'armv8b', 'armv8l']:
        return 'linux-arm64'
    elif system == 'darwin':
        return 'osx-arm64'
    elif system == 'linux':
        return 'linux-x64'
    else:
        return 'unknown'


runtime = get_runtime()
print(f"The script is running on: {runtime}")
``` 
6. Use runtime to select runtime specific subfolder
